api/src/main/java/net/countercraft/movecraft/events/SignTranslateEvent.java

```diff
@@ -2,46 +2,89 @@
 
 import net.countercraft.movecraft.MovecraftLocation;
 import net.countercraft.movecraft.craft.Craft;
+import net.countercraft.movecraft.sign.AbstractSignListener;
+import net.kyori.adventure.text.Component;
+import org.bukkit.block.BlockFace;
 import org.bukkit.event.HandlerList;
 import org.jetbrains.annotations.NotNull;
 
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
+// TODO: Rewrite to use the adventure API
 public class SignTranslateEvent extends CraftEvent{
     private static final HandlerList HANDLERS = new HandlerList();
     @NotNull private final List<MovecraftLocation> locations;
-    @NotNull private final String[] lines;
+    @NotNull private final AbstractSignListener.SignWrapper backing;
     private boolean updated = false;
 
+    @Deprecated(forRemoval = true)
     public SignTranslateEvent(@NotNull Craft craft, @NotNull String[] lines, @NotNull List<MovecraftLocation> locations) throws IndexOutOfBoundsException{
         super(craft);
         this.locations = locations;
-        if(lines.length!=4)
-            throw new IndexOutOfBoundsException();
-        this.lines=lines;
+        List<Component> components = new ArrayList<>();
+        for (String s : lines) {
+            components.add(Component.text(s));
+        }
+        this.backing = new AbstractSignListener.SignWrapper(null, components::get, components, components::set, BlockFace.SELF);
+    }
+
+    public SignTranslateEvent(@NotNull Craft craft, @NotNull AbstractSignListener.SignWrapper backing, @NotNull List<MovecraftLocation> locations) throws IndexOutOfBoundsException{
+        super(craft);
+        this.locations = locations;
+        this.backing = backing;
     }
 
     @NotNull
-    @Deprecated
+    @Deprecated(forRemoval = true)
     public String[] getLines() {
+        // TODO: Why does this set it to updated? This is just reading...
         this.updated = true;
-        return lines;
+        return backing.rawLines();
```

Comment:
Because the lines can be updated externally, and we needed all signs to be marked as updated for it to display properly to clients (see #671).

Reply:
I see, makes sense.

For the "sending directly to client" stuff, there is a separate "sendUpdatePacket" method that can be implemented